Library/Homebrew/api/formula.rb

```diff
@@ -20,6 +20,17 @@ def formula_api_path
         def fetch(name)
           Homebrew::API.fetch "#{formula_api_path}/#{name}.json"
         end
+
+        sig { returns(Array) }
+        def all_formulae
+          @all_formulae ||= begin
+            json_formulae = JSON.parse((HOMEBREW_CACHE_API/"#{formula_api_path}.json").read)
+
+            json_formulae.to_h do |json_formula|
+              [json_formula["name"], json_formula.except("name")]
+            end
+          end
```

Comment:
What actually downloads this?

Reply:
update.sh

Reply:
This was mainly for a few reasons:

* So that `brew fetch/install/upgrade/reinstall`, `brew tap` and `brew update/update-reset` remain the only commands to require an internet connection rather than every command requiring an internet connection.
* The speed of other commands remain fast, rather than slowed down by internet connections.
* The long-term goal is probably to do something like this anyway since we will likely want JSON to be the default for non-developers at least. Note that the update.sh downloading does not actually happen without `HOMEBREW_INSTALL_FROM_API` being set anyway.

Reply:
Yeh, I think the caching side of this is really valuable and having `update.sh` do downloads is too but that we should also add on-demand downloading so we're not reliant on auto-update for people to have up-to-date metadata.

Reply:
> so we're not reliant on auto-update for people to have up-to-date metadata.

Are we not already under git?

Reply:
@Bo98 We are but this change essentially maintains the Git workflow but just makes it faster by not having to `git fetch` homebrew/core.

Instead, we could have a workflow where we only automatically run `brew update` once a day/week and commands like `brew info` or `brew outdated` where people care more about having the most up-to-date data can actually pull it immediately.

This was the intent of the previous workflow. It's desirable to add caching on top of that but not move to `brew update`, a slow command, as the only way to update that metadata.

Reply:
Personally, I feel like it might be a good idea to have two levels for this. `HOMEBREW_INSTALL_FROM_API` can be used by those who don't mind the internet connectivity but just don't want to have the huge tap repos on their machine. We could also add `HOMEBREW_OFFLINE` or something like that which, when set, will mean that Homebrew never needs a connection unless explicitly requested through one of those commands.

Additionally, maybe a `--offline` flag on certain commands that will fall back on the cache when an internet connection would normally be used.

It seems to me that there are two reasons why one might want to use offline Homebrew: speed and so you can do certain things if you temporarily don't have an internet connection. For users who want to make sure that Homebrew is the fastest it can be, they may want to have an env variable. For those instances where you're temporarily disconnected, it might make more sense to have the `--offline` flag.

Reply:
I've been thinking about this a lot but ignoring speed of execution etc, there's still a couple concerns for me that I feel need to be addressed first:

* Consistency between API and non-API behaviour

Despite the title, the main point of this PR was to improve the consistency of behaviour between API and non-API code paths by effectively consolodating them and making it largely unnecessary for individual commands to have any API-specific logic.

But there's two sides to consistency: consistency in terms of the code but also consistency in terms of behaviour seen on the user's side.

Currently, `brew update` is responsible for updating locally stored version information. To take an example: `brew outdated` only works when you have first run `brew update`, otherwise it checks based on when you last ran `brew update`. A change to make it auto-fetch every time it runs may indeed be desirable, but I don't think having varying behaviour based on each tap is however.

Right now there's a very defined path (similar to `apt`) where you update your locally stored version information, and then run commands that operate on that information. It's fine to change this (and agree this can be desirable to) but to have a situation where "these formulae behave like this but these other formulae don't" is a worse situation IMO as it just confuses users more, and any benefit of "not needing to `brew update`" is lost as it'll just not be true for anything outside Homebrew/core.

* Formula migrations

We need to have a system to handle formulae migrations - both tap migrations and renames. This is actually quite important as we do this ourselves several times a year.

Currently this lives in `update-report`. This makes sense because it only needs to be done after updating taps, and it also needs to know the previous state of the tap as it does a diff to find deleted formulae. We'll need to consider how this works if tap updates no longer live in `brew update`.

Reply:
If it helps: I don't consider the desired outcome as being "`brew update` never needs to be run" but instead "`brew update` only runs automatically once a day/week instead of once every 5m by default" and "for users not using taps: we can avoid running `brew update` in more cases".

`brew update` being auto-run and the 5m default time predates Homebrew/brew having tags and was mainly to avoid people repeatedly reporting issues with formulae that have since been updated. We'll continue to use `brew update` for the cases you mention but it seems actively desirable for us to enable `brew outdated` and `brew info` and friends to be able to always get up-to-date information.

As above, on-disk caching if you don't have an internet connection is desirable but most users are using Homebrew most of the time with an internet connection and interacting with formulae from homebrew/core so let's optimise for that use-case.

Reply:
I'm torn about this and could be persuaded to go either way. Here's my gut response:

While consistency is important, `HOMEBREW_INSTALL_FROM_API` is already intentionally inconsistent. Part of the point is that there won't be issues like those that arise when people don't run `brew update`. It's also meant to make using Homebrew smoother and easier to use, and part of that is making it more intuitive. Since we don't support installing old versions of formulae, I think it's reasonable to automatically use the latest (and only supported) version. I'm not saying it's not inconsistent, but it feels like it makes Homebrew more intuitive to use for most people. And, since this feature is intentionally scoped to homebrew/core and homebrew/cask, it seems reasonable that those would operate slightly differently from other taps.

So, updating this file more frequently than on every `brew update` call seems like the choice that will make using Homebrew the easiest for most people.

Thanks for pointing out the issue with formula/tap migrations, though. That will need to be addressed if we take this route.

Of course, those who are more advanced users and might want the extra flexibility can always opt-out (or not opt-in). If we're concerned, we could also have an in-between state where you opt-in to using the API but not to auto-updates.

Alternatively, as a compromise, we could probably make it so it only actually re-downloads the file for certain commands (e.g. `install`, `outdated`, `upgrade`, etc) and will just use the cache from `update` for other commands where it's not as important to have super up-to-date info.

Reply:
>  Part of the point is that there won't be issues like those that arise when people don't run brew update. It's also meant to make using Homebrew smoother and easier to use, and part of that is making it more intuitive.

This is fine. The problem I have is telling users to change how they use Homebrew depending on whether taps are Git based or API based. Most people will not understand that. People will be confused when they see `brew outdated` "work" for Homebrew/core formulae but not for other taps.

> I'm not saying it's not inconsistent, but it feels like it makes Homebrew more intuitive to use for most people. And, since this feature is intentionally scoped to homebrew/core and homebrew/cask, it seems reasonable that those would operate slightly differently from other taps.

In general for making `brew update` not required: yes, I agree.
For making behaviour vary depending on each formula: no, I think this is more confusing.

Reply:
> While consistency is important, HOMEBREW_INSTALL_FROM_API is already intentionally inconsistent. Part of the point is that there won't be issues like those that arise when people don't run brew update. It's also meant to make using Homebrew smoother and easier to use, and part of that is making it more intuitive. Since we don't support installing old versions of formulae, I think it's reasonable to automatically use the latest (and only supported) version. I'm not saying it's not inconsistent, but it feels like it makes Homebrew more intuitive to use for most people. And, since this feature is intentionally scoped to homebrew/core and homebrew/cask, it seems reasonable that those would operate slightly differently from other taps.

I agree.

> Alternatively, as a compromise, we could probably make it so it only actually re-downloads the file for certain commands (e.g. install, outdated, upgrade, etc) and will just use the cache from update for other commands where it's not as important to have super up-to-date info.

Also agreed. Even in those cases, though, if you try to `brew install` but don't have an internet connection it should (eventually) probably try to download the latest version, output a warning and continue with the installation using the cached version.

> This is fine. The problem I have is telling users to change how they use Homebrew depending on whether taps are Git based or API based. Most people will not understand that. People will be confused when they see brew outdated "work" for Homebrew/core formulae but not for other taps.

Most people already don't rely on third-party taps and rely on autoupdate rather than an explicit `brew update`.

> In general for making brew update not required: yes, I agree.

Cool, good.

> For making behaviour vary depending on each formula: no, I think this is more confusing.

I don't think it's varying per-formula, I think it's just that we have the ability to provide up-to-date information without a `brew update` (automatic or otherwise) for various operations (we can debate exactly which) through the API.

Eventually this will be something we can offer for taps too.

Reply:
> Most people already don't rely on third-party taps

I do not think we should see third-party taps as some advanced use case. A few formulae from third party taps do make up our top 100 most installed formulae.

The concept of a tap can be mostly invisible to the user. Any novice user could [follow instructions](https://devcenter.heroku.com/articles/getting-started-with-nodejs?singlepage=true#set-up) to run `brew install heroku/brew/heroku` and to them it is like installing any other formula.

> I don't think it's varying per-formula

From a technical stance, sure it's really per-tap. From a user experience point of view, the example above will appear on the surface that the formula `heroku/brew/heroku` behaves unintuitively or "broken".

Reply:
Could we help combat this problem by, maybe, always auto-updating a tap that's explicitly referenced in a command? Meaning if someone runs `brew install heroku/brew/heroku` or even, maybe, `brew info heroku` where the formula is known to be part of a third-party tap, we automatically fetch only that tap? Or maybe check to see if a newer version of the formula file is available (if the tap is hosted on GitHub) and if so, warn the user about it?

Or if not that exact solution, are there other things we can do to improve the consistency?

Reply:
> From a technical stance, sure it's really per-tap. From a user experience point of view, the example above will appear on the surface that the formula heroku/brew/heroku behaves unintuitively or "broken".
> 
> If this variance of behaviour is what we want, then we should have various signposts to cover the shortcomings in user experience.

If this happens: we could fix it. I don't think we want a lowest-common denominator experience for homebrew/core, though, just because tap behaviour may be inconsistent.

I'm also not convinced that "consistency" is actually always preferable here.

> Meaning if someone runs brew install heroku/brew/heroku or even, maybe, brew info heroku where the formula is known to be part of a third-party tap, we automatically fetch only that tap?

The `info` case: I don't think we should do. The install case: we should and it would be fairly easy to implement (remembering that this decision is in Bash not Ruby).

> Or if not that exact solution, are there other things we can do to improve the consistency?

Personally, I don't think it's worth getting too hung up on consistency yet.

Reply:
> I don't think we want a lowest-common denominator experience for homebrew/core, though, just because tap behaviour may be inconsistent.

Basically the point I have is to not forget user experience in among the technical changes. While it might not be a problem at this particular second - it is important to however address that before we can consider making this a default and it may have a impact on the design approaches we take at this stage.

> I'm also not convinced that "consistency" is actually always preferable here.

Consistency is usually easiest to understand. It's of course not the only way forward, but where there's inconsistency it needs to be easily understood and obvious without having to teach people what a tap, Git or an API is.

> Meaning if someone runs brew install heroku/brew/heroku or even, maybe, brew info heroku where the formula is known to be part of a third-party tap, we automatically fetch only that tap?

Also remember the `brew outdated` case where the command doesn't include any particular formula. We may need to do things like checking if taps themselves are outdated. There's a few cases to consider.

Reply:
> `brew outdated` behaviour is a perfect example here. It says "nothing is outdated", I run brew upgrade and things are updated anyway. This is not consistent or understandable behaviour unless you happen to know that `brew outdated` doesn't auto-update and `brew upgrade` does.

Oh to be clear: I agree with that. But this issue will still exist for third-party formulae so it's still important to improve the experience for them too, hence the potential for an additional tap outdated check (unless it's too slow).

Reply:
> But this issue will still exist for third-party formulae

Yeh, I guess my take is: it being fixed for homebrew/core is a better situation than it being broken/weird everywhere but being consistent and this isn't something we need to fix for the V1 of rolling out the API to end-users.

Reply:
I think what we're settling is me revisiting this later on and proposing PRs where I feel the UX could be better.

Reply:
100% 👍🏻